src/LogicTasks/Semantics/Pick.hs

```diff
@@ -14,26 +15,36 @@ import Control.Monad.Output (
   translate,
   )
 
-import Test.QuickCheck (Gen, elements, vectorOf)
+import Test.QuickCheck (Gen, elements, vectorOf, suchThat)
 
-import Config (BaseConfig(..), CnfConfig(..), Number(..), PickConfig(..), PickInst(..))
-import Formula.Util (mkCnf, isSemanticEqual)
-import Formula.Types (availableLetter, genCnf, getTable, literals)
+import Config (Number(..), PickConfig(..), PickInst(..))
+import Formula.Util (isSemanticEqual)
+import Formula.Types (availableLetter, getTable, literals)
 import Formula.Printing (showIndexedList)
-import Util (checkCnfConf, tryGen)
 import LogicTasks.Helpers (example, extra)
 import Control.Monad (when)
 import Data.Maybe (fromJust)
-import Data.List (nubBy)
-
+import Data.List (nubBy, sort)
+import Trees.Generate (genSynTree)
+import Tasks.SynTree.Config (checkSynTreeConfig, SynTreeConfig (..))
+import Trees.Print (display)
+import Data.List.Extra (nubOrd)
+import Trees.Helpers (collectLeaves)
+import Trees.Formula ()
 
 genPickInst :: PickConfig -> Gen PickInst
-genPickInst PickConfig{ cnfConf = CnfConfig {baseConf = BaseConfig{..}, ..}, ..} = do
-    cnfs <- tryGen (vectorOf amountOfOptions (getCnf usedLiterals)) 100 ((amountOfOptions ==) . length . nubBy isSemanticEqual)
+genPickInst PickConfig{..} = do
+    trees <- vectorOf amountOfOptions (genSynTree syntaxTreeConfig) `suchThat` \trees ->
+      length (nubBy isSemanticEqual trees) == amountOfOptions &&
+      length (nubOrd (map (nubOrd . sort . collectLeaves) trees)) == 1
```

Comment:
This line seems to be a bit problematic (non-termination), even though the constraint should be fine. Only checking for the same amount of atomics does not result in non-termination. So I guess there is a problem with the generation of trees having the exact same atomics.

Reply:
Without having looked at the code at all, and a bit on the run, here is a thought:

If you need to generate n trees all having the exact same atoms, then maybe instead of:
- generating n trees quite randomly and just hoping that often enough they will have the same sets of atoms

you could do:
- generate n tree shapes (`Tree BinOp ()`) randomly, and separately generate the set of atoms you want them all to agree on, and then from that set generate random permutations that you can use to decorate the `()`-leafs in your previously obtained shapes

A strategy similar to that should already be present somewhere else in a task type.